torch_xla/csrc/tensor_methods.cpp

```diff
@@ -385,6 +385,49 @@ XLATensor XLATensor::get_dimensions_size(const XLATensor& input,
                           at::ScalarType::Int);
 }
 
+void XLATensor::sgd_optimizer_step(const XLATensor& found_inf, XLATensor& step,
+                                   XLATensor& param, const XLATensor& d_p,
+                                   XLATensor& buf, double weight_decay,
+                                   double momentum, double lr, double dampening,
+                                   bool nesterov) {
+  /* XLA version of sgd algorithm
+   * https://github.com/pytorch/pytorch/blob/master/torch/optim/_functional.py#L162-L180
+   */
+  auto d_p_value = d_p.GetIrValue();
+  // weight_decay
+  if (weight_decay != 0) {
+    ir::Value weight_decay_value =
+        GetIrValueForScalar(weight_decay, param.shape(), param.GetDevice());
+    d_p_value = d_p_value + param.GetIrValue() * weight_decay_value;
+  }
+  // momentum
+  if (momentum != 0) {
+    auto buf_value = buf.GetIrValue();
+    ir::Value momentum_value =
+        GetIrValueForScalar(momentum, param.shape(), param.GetDevice());
+    ir::Value dampening_factor =
+        GetIrValueForScalar(1.0 - dampening, param.shape(), param.GetDevice());
+    buf_value = ir::ops::Where(
+        step.GetIrValue(),
+        buf_value * momentum_value + d_p_value * dampening_factor, d_p_value);
+    d_p_value = nesterov ? d_p_value + buf_value * momentum_value : buf_value;
+    buf.SetInPlaceIrValue(
+        ir::ops::Where(found_inf.GetIrValue(), buf.GetIrValue(), buf_value));
+  }
+  // update param
+  ir::Value lr_value =
+      GetIrValueForScalar(lr, param.shape(), param.GetDevice());
+  param.SetInPlaceIrValue(
+      ir::ops::Where(found_inf.GetIrValue(), param.GetIrValue(),
+                     param.GetIrValue() - d_p_value * lr_value));
+  // update step counter
+  ir::Value one_value =
+      GetIrValueForScalar(1.0, step.shape(), step.GetDevice());
+  step.SetInPlaceIrValue(ir::ops::Where(found_inf.GetIrValue(),
+                                        step.GetIrValue(),
+                                        step.GetIrValue() + one_value));
+}
```

Comment:
I think we should move this logic to a new `sgd_optimizer_step` Node under `torch_xla/csrc/ops/sgd_optimizer_step.h` and `torch_xla/csrc/ops/sgd_optimizer_step.cpp`. There are 2 reasons

1. It is more clear when user dump the IR graph this way. They will see a `xla::sgd_optimizer_step` in the graph instead of a bunch of `add`, `where` etc
2. We are trying to move all computation logic in the IR level (tensor_method.cpp) either to pytorch dispatcher or XLA lowering in the process of Lazy Tenor Upstream. 

You can use [logsumexp](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/ops/logsumexp.h) Node as an example.

Reply:
I think a better example would be [collectivepermute](https://github.com/pytorch/xla/commit/39d22ee6ee6af6b022ba90b72cc2cca071316136), since it is also a op that doesn't exist from pytorch's perspective.

Reply:
Thanks for the pointer, I've updated the PR with the following changes:
1. Incorporate `amp.syncfree.GradScaler` into the original `amp.GradScaler`.
2. Make `amp.syncfree.SGD` a subclass of `torch.optim.SGD`.
3. Add the new `SgdOptimizerStep` Node.